src/assembler.jl

```diff
@@ -164,22 +166,33 @@ stiffness (and force) in `A`, given the element degrees of freedom `dofs`.
 This is equivalent to `K[dofs, dofs] += Ke` and `f[dofs] += fe`, where `K` is the global
 stiffness matrix and `f` the global force/residual vector, but more efficient.
 """
-assemble!(::AbstractSparseAssembler, ::AbstractVector{Int}, ::AbstractMatrix, ::AbstractVector)
+assemble!(::AbstractSparseAssembler, ::AbstractVector{<:Integer}, ::AbstractMatrix, ::AbstractVector)
 
-@propagate_inbounds function assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{Int}, Ke::AbstractMatrix)
+@propagate_inbounds function assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{<:Integer}, Ke::AbstractMatrix)
     assemble!(A, dofs, Ke, eltype(Ke)[])
 end
-@propagate_inbounds function assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{Int}, fe::AbstractVector, Ke::AbstractMatrix)
+@propagate_inbounds function assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{<:Integer}, fe::AbstractVector, Ke::AbstractMatrix)
     assemble!(A, dofs, Ke, fe)
 end
-@propagate_inbounds function assemble!(A::AssemblerSparsityPattern, dofs::AbstractVector{Int}, Ke::AbstractMatrix, fe::AbstractVector)
+@propagate_inbounds function assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{<:Integer}, Ke::AbstractMatrix, fe::AbstractVector)
     _assemble!(A, dofs, Ke, fe, false)
 end
-@propagate_inbounds function assemble!(A::AssemblerSymmetricSparsityPattern, dofs::AbstractVector{Int}, Ke::AbstractMatrix, fe::AbstractVector)
+@propagate_inbounds function assemble!(A::SymmetricCSCAssembler, dofs::AbstractVector{<:Integer}, Ke::AbstractMatrix, fe::AbstractVector)
     _assemble!(A, dofs, Ke, fe, true)
 end
 
-@propagate_inbounds function _assemble!(A::AbstractSparseAssembler, dofs::AbstractVector{Int}, Ke::AbstractMatrix, fe::AbstractVector, sym::Bool)
+@propagate_inbounds function _sortdofs_for_assembly(A::AbstractSparseAssembler, dofs::AbstractVector)
+    ld = length(dofs)
+    permutation = A.permutation
+    sorteddofs = A.sorteddofs
```

Comment:
Here as well (but since you nicely extracted this function, could just dispatch on the union of the existing assemblers?)

Reply:
What exactly is the issue here and below?

Reply:
That the input is `AbstractSparseAssembler`, but specific fields are used (i.e. the implementer of an `AT <:  AbstractSparseAssembler` must provide these specific fields. 
In this case, I guess the function is specific to `Union{CSCAssembler, SymmetricCSCAssembler}`, and I would think it makes more sense to dispatch on that? 

A more general question perhaps for my understanding, what methods do you expect to redefine for a custom `AbstractSparseCSCAssembler`?

Reply:
I see it as follows: We provide sensible defaults for these dispatches and if you want to do something different, then you can do so.

Reply:
This is also the same for AbstractGrid and brings the advantage that you only need to dispatch the specific things that actually changed

Reply:
I guess we have different opinions here then :) I see that it makes it "easier" to hack, but direct field access interfaces instead of accessor functions makes renaming a field a breaking change. I now recall that this was discussed before, ref https://github.com/Ferrite-FEM/Ferrite.jl/pull/655#issuecomment-1490372711 regarding the Grid/DofHandler, but I don't think a consensus was found.

Reply:
IMO we should
* Dispatch `show`, `vector_handle`, and `_sortdofs_for_assembly` on `Union{SymCSCAssembler, CSCAssembler}`
* Use `vector_handle` in `_assemble!`
So that if you bring your own `assembler::AbstractCSCAssembler` that should use `_assemble!`, you must define `vector_handle`, `matrix_handle`, and `_sortdofs_for_assembly`

Reply:
A good example of why we prefer this is it given in FerriteDistributed
https://github.com/Ferrite-FEM/FerriteDistributed.jl/blob/main/src/NODDofHandler.jl#L31-L108
https://github.com/Ferrite-FEM/FerriteDistributed.jl/blob/main/src/NODDofHandler.jl#L127-L143
Although it is not as extreme for the assembler

Reply:
AFAIU most of those won't work with https://github.com/Ferrite-FEM/Ferrite.jl/commit/7e8a57150738094bb951d6e672fdeed205d0a1ff due to the subdofhandler, and thus wouldn't be redundant? Which is my point, that this approach quickly leads to breaking the code. 

But just to be clear - I strongly support writing our functions to support abstract types. I just think there should be functions then and not field access, because functions are much easier to overload / replace when changes are made.

Reply:
Just another example, if we would consistently rewrite the `AbstractGrid` dispatches to use accessor functions, then those https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/src/Grid/grid.jl#L385-L467 would need to be dispatched for https://github.com/Ferrite-FEM/Ferrite.jl/blob/731dfbb0f0733d943de90a55c0a7c31b82704d2e/src/Adaptivity/ncgrid.jl which only adds a lot of redundant code. I think this issue was discussed back then when we started to use some other grid types. We also have in the docs the statement that one can rely on the fallback `AbstractGrid` dispatches in case specific parts of the datastructure don't change: https://ferrite-fem.github.io/Ferrite.jl/stable/manual/grid/#AbstractGrid

Reply:
I started generalizing this in https://github.com/Ferrite-FEM/Ferrite.jl/pull/662/files#diff-13ae190cb008d749905405bddca2a9ce395b303e12c781186df0d908efd05a90. 
I would expect that you should only need to define 
```julia
get_coordinate_type
getcells
getnodes
getcellsets
getnodesets
getfacesets
getedgesets
getvertexsets
```
To make an equivalent grid as the current, potentially with more features such as the ncgrid. 

> We also have in the docs the statement that one can rely on the fallback AbstractGrid dispatches in case specific parts of the datastructure don't change: https://ferrite-fem.github.io/Ferrite.jl/stable/manual/grid/#AbstractGrid

"In case that certain structures are preserved from the Ferrite.Grid type, you don't need to dispatch on your own type, but rather rely on the fallback AbstractGrid dispatch."

Oh, I was unaware of this statement. This is a pretty strong statement in the docs, which will mean any internal change to the grid has to be considered a breaking change. Do we really want to promise this with 1.0? (I mean, even if it works and we decide to keep allowing it, I think we should add a disclaimer to that statement that such changes will not be considered breaking)

Reply:
> 
> ```julia
> get_coordinate_type
> getcells
> getnodes
> getcellsets
> getnodesets
> getfacesets
> getedgesets
> getvertexsets
> ```
> 
> To make an equivalent grid as the current, potentially with more features such as the ncgrid.

Yeah that's the point, I don't need to define any of those, since the structure is preserved for those functions. https://github.com/Ferrite-FEM/Ferrite.jl/blob/731dfbb0f0733d943de90a55c0a7c31b82704d2e/src/Adaptivity/ncgrid.jl is the complete implementation of the datastructure, there is no additional code to it, which wouldn't be the case if we use the approach of: https://github.com/Ferrite-FEM/Ferrite.jl/pull/662/files#diff-13ae190cb008d749905405bddca2a9ce395b303e12c781186df0d908efd05a90

> Do we really want to promise this with 1.0? (I mean, even if it works and we decide to keep allowing it, I think we should add a disclaimer to that statement that such changes will not be considered breaking)

Disclaimer is fine for me. I'm not sure if either @KristofferC or @fredrikekre convinced me back then that the fallback dispatches are a good idea or maybe I'm mixing up memories

Reply:
> AFAIU most of those won't work with [7e8a571](https://github.com/Ferrite-FEM/Ferrite.jl/commit/7e8a57150738094bb951d6e672fdeed205d0a1ff) due to the subdofhandler, and thus wouldn't be redundant? Which is my point, that this approach quickly leads to breaking the code.
> 
> But just to be clear - I strongly support writing our functions to support abstract types. I just think there should be functions then and not field access, because functions are much easier to overload / replace when changes are made.

(Wrong commit?) But okay, let us discuss API design then for a moment, because this is what this comment (and linked ones) boil down to. 

You are right, breaking internals is breaking for subpackages which depend on the internals. And there is no good around it. Every change, even a simple bug fix, can be potentially breaking. My point is that the subpackages like for example FerriteDistributed should not diverge from the Ferrite codebase. We changed the features for the dof handler and the distributed dof handler is not updated yet. So if FerriteDistributed would be registered and we would tag a new release of Ferrite with the breaking internal change, then the current release of FerriteDistributed it would be simply not compatible with the release. Users will probably expect that if they pull the new Ferrite release with FerriteDistributed, that the new features will integrate, which we cannot guarantee in general. In the worst case this will lead to a silent failure.

To give another example: The node ordering for our elements is technically an internal invariant and should not matter for the user. However, changing it breaks the mesh readers and exporters. Simply because we need to rely on some internal assumptions. 

I think this makes the point clear that breaking internals will require checking downstream packages anyway if we rely in one way or another on internal details.

However, in software development the pure renaming of member variables almost never happens. Furthermore, I would argue that it is sensible to assume that specific sets of similar data types share similar members within a framework, especially if they do essentially the same thing in a slightly different way which we cannot easily express in the algorithms.

Reply:
(The commit link was copied from FerriteDistributed.jl's readme stating which Ferrite version it should work with)

Of course this doesn't prevent issues with downstream packages, and as you say, there isn't an easy way around. My point was just that having these functions can reduce the risk somewhat (and make fixing things easier because it encourages using these functions more), at IMO a low cost of duplicating functions accessing the fields. One example of a typical rename would be to put part of the functionality in a sub-struct as we did for `CellValues` recently. 

But at the end of the day, this is a design choice. I commented because I thought this was just a mistake, but if not and it can be argued then for the assembler this is a minor issue.